agent/agent.go

```diff
@@ -241,15 +241,43 @@ func (a *Agent) Start() error {
 		return fmt.Errorf("Failed to setup node ID: %v", err)
 	}
 
+	// create the config for the rpc server/client
+	consulCfg, err := a.consulConfig()
+	if err != nil {
+		return err
+	}
+
 	// Setup either the client or the server.
 	if c.Server {
-		server, err := a.makeServer()
+		serverUp := make(chan int)
+		consulCfg.ServerUp = func() { close(serverUp) }
+
+		server, err := consul.NewServerLogger(consulCfg, a.logger)
 		if err != nil {
-			return err
+			return fmt.Errorf("Failed to start Consul server: %v", err)
 		}
 
 		a.delegate = server
-		a.state.Init(c, a.logger, server)
+
+		// wait for the server to be ready
+		// then trigger an anti-entropy run ???
+
+		// this needs review since there is no for the state to be
```

Comment:
I ended up going back to the callback approach since that was simpler with the non-blocking write to the channel. Otherwise, you'd need two channels and a go routine: one for the notifications on the producer side and one on the consumer side which handles the buffering logic. It is possible that I broke this earlier and just fixed it again. The change also contains other cleanups in the initialization of the localState so that looks better in general.

See 1d6c9c24